app/components/MultiStepForm.tsx

```diff
@@ -0,0 +1,192 @@
+     */
+    validate?: ValidateValues<F>;
+    onDone(values: F): void;
+    onPageActive?(step: keyof F, values: Partial<F>): void;
+    /**
+     * Pages of the transaction flow declared as a mapping of components to corresponding substate.
+     * Declaration order defines the order the pages are shown.
+     */
+    children: OrRenderValues<F, FormChildren<F>>;
+}
+
+interface InternalValueStoreProps<F extends Record<string, unknown>>
+    extends Props<F> {
+    initialValues?: F;
+}
+
+interface ExternalValueStoreProps<F extends Record<string, unknown>>
+    extends Props<F> {
+    valueStore: [Partial<F>, Dispatch<SetStateAction<Partial<F>>>];
+}
+
+export type MultiStepFormProps<F extends Record<string, unknown>> =
+    | InternalValueStoreProps<F>
+    | ExternalValueStoreProps<F>;
+
+export default function MultiStepForm<
+    // eslint-disable-next-line @typescript-eslint/no-explicit-any
+    F extends Record<string, any>
+>(props: MultiStepFormProps<F>) {
+    const { path: matchedPath } = useRouteMatch();
+    const {
+        children,
+        validate = () => undefined,
+        onDone,
+        onPageActive = noOp,
+        baseRoute = matchedPath,
+    } = props;
+    const initialValues =
+        (props as InternalValueStoreProps<F>).initialValues ?? ({} as F);
+    const { pathname } = useLocation();
+    const internalValueStore = useState<Partial<F>>(initialValues);
+    const externalValueStore = (props as ExternalValueStoreProps<F>).valueStore;
+    const [values, setValues] = externalValueStore ?? internalValueStore;
+    const dispatch = useDispatch();
+
+    const getChildren = useCallback(
+        (v: Partial<F>) =>
+            typeof children === 'function' ? children(v) : children,
+        [children]
+    );
+
+    const pages = useMemo(
+        () => makeFormPageObjects(baseRoute, getChildren(values)).reverse(),
+        [getChildren, values, baseRoute]
+    );
+
+    const currentPage = pages.find((p) => pathname.startsWith(p.route));
+
+    useEffect(() => {
+        if (currentPage?.substate) {
+            onPageActive(currentPage?.substate, values);
+        }
+        // eslint-disable-next-line react-hooks/exhaustive-deps
+    }, [currentPage?.substate]);
+
+    useUpdateEffect(() => {
+        throw new Error(
+            'Changing value store during the lifetime of MultiStepForm will result in errors.'
+        );
+    }, [externalValueStore === undefined]);
+
+    const handleNext = (substate: keyof F) => (v: Partial<F>) => {
+        const newValues = { ...values, [substate]: v };
+        setValues(newValues);
+
+        const newPages = makeFormPageObjects(baseRoute, getChildren(newValues));
+        const currentIndex = newPages.findIndex((p) => p.substate === substate);
+
+        if (currentIndex === -1) {
+            // Could not find current page. Should not happen.
+            dispatch(replace(baseRoute));
```

Comment:
if this is an exceptional situation, maybe it should use an exception? :upside_down_face:

Reply:
I don't agree.. an unforeseeable situation might as well take the user back to a place where he/she can start over. It would make sense to add some logging, but then again, it would only make sense to log if we could track the users path to where the exceptional situation occurred ;)

Reply:
sure, lets add a "TODO" for logging then?